hash_reader.go

```diff
@@ -33,98 +32,138 @@ type HashReader interface {
 }
 
 type hashReader struct {
-	ionReader      ion.Reader
-	hasherProvider IonHasherProvider
+	ionReader ion.Reader
+	hasher    hasher
+	typeIon   ion.Type
 }
 
 func NewHashReader(ionReader ion.Reader, hasherProvider IonHasherProvider) HashReader {
-	hashReader := &hashReader{ionReader, hasherProvider}
+	hashReader := &hashReader{ionReader, *newHasher(hasherProvider), ion.NoType}
 
 	return hashReader
 }
 
 func (hashReader *hashReader) SymbolTable() ion.SymbolTable {
-	panic("implement me")
 	return hashReader.ionReader.SymbolTable()
 }
 
 func (hashReader *hashReader) Next() bool {
-	panic("implement me")
+	switch hashReader.typeIon {
```

Comment:
This is incorrect logic.

There are 2 conditions (Ion container and Ion scalar types)

Reply:
In dot net, ionType is defined as

public enum IonType : short
    {
        None = -1,
        Null = 0,
        Bool = 1,
        Int = 2,
        Float = 4,
        Decimal = 5,
        Timestamp = 6,
        Symbol = 7,
        String = 8,
        Clob = 9,
        Blob = 10,
        List = 11,
        Sexp = 12,
        Struct = 13,
        Datagram = 14,
    }

ionType.IsContainer() is defined as
public static bool IsContainer(this IonType t) => t >= IonType.List;

The first part of my switch statement handles the container types (with the exception of Datagram because that's not defined in ion-go)

ionType.IsScalar() is defined as
public static bool IsScalar(this IonType t) => t > IonType.None && t < IonType.List;

The second part of my switch statement handles the scalar types.

Since there is no overlap between container and scalar types, we can combine this logic into just one switch statement.

Reply:
This function is missing a step: setting the typeIon

Reply:
Can you create github issue on ion-go to create functions: isScalar(), isContainer() within type.go so this switch statement can be abstracted out. I am pretty sure we will need to call this again in the serializer functions

Reply:
> This function is missing a step: setting the typeIon

Done.

Reply:
> Can you create github issue on ion-go to create functions: isScalar(), isContainer() within type.go so this switch statement can be abstracted out. I am pretty sure we will need to call this again in the serializer functions

Done. https://github.com/amzn/ion-go/issues/34